File: apps/mobile/modules/synq-nearby/ios/SynqNearbyModule.swift
import ExpoModulesCore
import CoreBluetooth

public final class SynqNearbyModule: Module, CBCentralManagerDelegate, CBPeripheralManagerDelegate {
  private let serviceUUID = CBUUID(string: "A7C10001-86B4-4E0A-9CA1-6D88F23A1B31")
  private var central: CBCentralManager?
  private var peripheral: CBPeripheralManager?
  private var token = ""

  public func definition() -> ModuleDefinition {
    Name("SynqNearby")
    Events("onEncounter", "onState")
    AsyncFunction("permissionState") { () -> String in
      switch CBManager.authorization { case .allowedAlways: return "granted"; case .denied, .restricted: return "denied"; default: return "undetermined" }
    }
    AsyncFunction("start") { (epochToken: String) in
      guard epochToken.utf8.count > 0 && epochToken.utf8.count <= 4 else { throw InvalidTokenException() }
      self.token = epochToken
      self.central = CBCentralManager(delegate: self, queue: nil)
      self.peripheral = CBPeripheralManager(delegate: self, queue: nil)
    }
    AsyncFunction("stop") { self.central?.stopScan(); self.peripheral?.stopAdvertising(); self.central=nil; self.peripheral=nil }
    OnAppEntersBackground { self.sendEvent("onState", ["state":"degraded-background"]) }
  }
  public func centralManagerDidUpdateState(_ central: CBCentralManager) {
    guard central.state == .poweredOn else { sendEvent("onState", ["state":"bluetooth-unavailable"]); return }
    central.scanForPeripherals(withServices: nil, options: [CBCentralManagerScanOptionAllowDuplicatesKey:false])
  }
  public func centralManager(_ central: CBCentralManager, didDiscover peripheral: CBPeripheral, advertisementData: [String:Any], rssi RSSI: NSNumber) {
    guard let data=advertisementData[CBAdvertisementDataServiceDataKey] as? [CBUUID:Data], let payload=data[serviceUUID], let observed=String(data:payload,encoding:.utf8), observed != token else { return }
    sendEvent("onEncounter", ["epochToken":observed,"rssiBucket":RSSI.intValue > -65 ? "likely-near":"uncertain","observedAt":ISO8601DateFormatter().string(from:Date())])
  }
  public func peripheralManagerDidUpdateState(_ peripheral: CBPeripheralManager) {
    guard peripheral.state == .poweredOn else { return }
    peripheral.startAdvertising([CBAdvertisementDataServiceDataKey:[serviceUUID:Data(token.utf8)]])
  }
}
private class InvalidTokenException: Exception { override var reason: String { "Epoch token must be 1 to 4 UTF-8 bytes" } }

File: apps/mobile/modules/synq-nearby/android/src/main/java/expo/modules/synqnearby/SynqNearbyModule.kt
package expo.modules.synqnearby
import android.Manifest
import android.bluetooth.*
import android.bluetooth.le.*
import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import android.os.ParcelUuid
import androidx.core.content.ContextCompat
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.util.UUID

class SynqNearbyModule: Module() {
 private val uuid=ParcelUuid(UUID.fromString("A7C10001-86B4-4E0A-9CA1-6D88F23A1B31")); private var token=""; private var callback:ScanCallback?=null
 override fun definition()=ModuleDefinition {
  Name("SynqNearby"); Events("onEncounter","onState")
  AsyncFunction("permissionState") { if(Build.VERSION.SDK_INT<31 || listOf(Manifest.permission.BLUETOOTH_SCAN,Manifest.permission.BLUETOOTH_ADVERTISE,Manifest.permission.BLUETOOTH_CONNECT).all{ContextCompat.checkSelfPermission(context,it)==PackageManager.PERMISSION_GRANTED}) "granted" else "undetermined" }
  AsyncFunction("start") { epochToken:String -> require(epochToken.toByteArray().size<=20);token=epochToken;startBle() }
  AsyncFunction("stop") { stopBle() }
  OnDestroy { stopBle() }
 }
 private val context get()=appContext.reactContext ?: throw IllegalStateException("No React context")
 private fun adapter()=(context.getSystemService(Context.BLUETOOTH_SERVICE) as BluetoothManager).adapter
 private fun startBle(){val a=adapter();if(!a.isEnabled){sendEvent("onState",mapOf("state" to "bluetooth-unavailable"));return};val data=token.toByteArray();a.bluetoothLeAdvertiser?.startAdvertising(AdvertiseSettings.Builder().setAdvertiseMode(AdvertiseSettings.ADVERTISE_MODE_LOW_POWER).setConnectable(false).build(),AdvertiseData.Builder().addServiceUuid(uuid).addServiceData(uuid,data).build(),object:AdvertiseCallback(){});callback=object:ScanCallback(){override fun onScanResult(t:Int,r:ScanResult){val found=r.scanRecord?.getServiceData(uuid)?.toString(Charsets.UTF_8)?:return;if(found==token)return;sendEvent("onEncounter",mapOf("epochToken" to found,"rssiBucket" to if(r.rssi>-65)"likely-near" else "uncertain","observedAt" to System.currentTimeMillis()))}};a.bluetoothLeScanner?.startScan(listOf(ScanFilter.Builder().setServiceUuid(uuid).build()),ScanSettings.Builder().setScanMode(ScanSettings.SCAN_MODE_LOW_POWER).build(),callback)}
 private fun stopBle(){val a=adapter();callback?.let{a.bluetoothLeScanner?.stopScan(it)};a.bluetoothLeAdvertiser?.stopAdvertising(object:AdvertiseCallback(){});callback=null}
}
